packages/hub/services/nonce-lock.ts

```diff
@@ -0,0 +1,79 @@
+import { inject } from '@cardstack/di';
+import BN from 'bn.js';
+import config from 'config';
+import { addSeconds } from 'date-fns';
+import shortUUID from 'short-uuid';
+import { nowUtc } from '../utils/dates';
+
+export default class NonceLock {
+  prismaManager = inject('prisma-manager', { as: 'prismaManager' });
+  ethersProvider = inject('ethers-provider', { as: 'ethersProvider' });
+
+  private async getNonce(accountAddress: string, chainId: number) {
+    let prisma = await this.prismaManager.getClient();
+    return await prisma.$transaction(
+      async (tx) => {
+        await tx.$executeRaw`SELECT * FROM account_nonces WHERE account_address = ${accountAddress} AND chain_id = ${chainId} FOR UPDATE`;
+
+        let accountNonce = await tx.accountNonce.findUnique({
+          where: { accountAddress_chainId: { accountAddress, chainId } },
+        });
+        let one = new BN(1);
+        let nextNonce = accountNonce?.nonce ? new BN(accountNonce.nonce.toString()).add(one) : new BN(0);
+        let nonceTTL = Number(config.get('nonceTTL'));
+        let now = nowUtc();
+        if (nextNonce.cmp(one) <= 0 || (accountNonce && addSeconds(accountNonce.updatedAt, nonceTTL) <= now)) {
+          let provider = this.ethersProvider.getInstance(chainId);
+          nextNonce = new BN(await provider.getTransactionCount(accountAddress, 'latest'));
+        }
+
+        if (accountNonce) {
+          await tx.accountNonce.update({
+            where: { id: accountNonce.id },
+            data: {
+              nonce: BigInt(nextNonce.toString()),
+              updatedAt: now,
+            },
+          });
+        } else {
+          await tx.accountNonce.create({
+            data: {
+              id: shortUUID.uuid(),
+              accountAddress,
+              chainId,
+              nonce: BigInt(nextNonce.toString()),
+              createdAt: now,
+              updatedAt: now,
+            },
+          });
+        }
+
+        return nextNonce;
+      },
+      { maxWait: 3000, timeout: 10000 }
+    );
+  }
+
+  async withNonce(accountAddress: string, chainId: number, cb: (nonce: BN) => Promise<any>) {
+    let prisma = await this.prismaManager.getClient();
+    let nonce = await this.getNonce(accountAddress, chainId);
```

Comment:
i think i see a problem here--you should actually block until the txn is submitted on chain. this assumes that the txn will succeed. Consider the fact that 2 txns call `withNonce` at exactly the same time. this will assign one of the txns a nonce and the other txn the next nonce. however, i think if the txn reverts, then you don't want to advance the nonce (probably we should look that up to be sure). you don't actually know that until you you send the txn. while the txn is being sent, i don't think you can assign a nonce, and any other calls to `withNonce` probably need to continue to block (via an `await`) until you know that the txn doesn't revert.

Reply:
I was implementing it like that, but then I thought the timeout of the database transaction would be too long because it needs to wait until the execution transaction publishes on the blockchain. That's why I put error handling to reset the nonce to zero in case the execution transaction is failed to publish (https://github.com/cardstack/cardstack/blob/e9ebe31f131a91481b81201ef254ff308dd9022d/packages/hub/services/nonce-lock.ts#L63).

But you're right. I'll include the execution transaction in that database transaction scope.

Reply:
updated